libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,102 @@
+#include "adc.h"
+#include "stm32f0xx.h"  
+
+#include <stdio.h>
+
+static uint8_t prv_get_channel(GPIOAddress *address) {
+  uint8_t adc_channel = address->pin;
+
+  switch (address->port) {
+     case GPIO_PORT_A:
+       if (address->pin > 7) { return 11; }
+       break;
+
+     case GPIO_PORT_B:
+       if (address->pin > 1) { return 11; }
+       adc_channel += 8;
+       break;
+
+     case GPIO_PORT_C:
+       if (address->pin > 5) { return 11; }
+       adc_channel += 10;
+       break;
+
+     default:
+       return 11;
+  }
+
+  return adc_channel;
+}
+
+void adc_init(ADCMode adc_mode) {
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    ADC_Resolution_12b,
+    adc_mode,
+    ADC_ExternalTrigConvEdge_None,
+    ADC_ExternalTrigConv_T1_TRGO,
+    ADC_DataAlign_Right,
+    ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  ADC_GetCalibrationFactor(ADC1);
+
+  // Calculate the ADC calibration factor 
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  // Continuous conversions must be initiated by setting the ADSTART 
+  if ((ADC1->CFGR1 >> 13) & 1) {
+    ADC_StartOfConversion(ADC1);
+  }
+}
+
+StatusCode adc_init_pin(GPIOAddress *address, ADCSampleRate adc_sample_rate) {
```

Comment:
They don't need to be configured for analog. I thought that support for individual sampling rates might have been a useful feature to have at the beginning, but I don't think that'll be the case at this point. I'll probably get rid of it.